Cloud Command System Migration

Aurelien30000 opened the pull request:
# Introduction

**This is the first pass of the cloud command migration for FAVS.** There will be a second one to restore old-fashioned command syntax, tracked as https://github.com/IntellectualSites/fastasyncvoxelsniper/issues/81.

_I have to highlight that this is my first experience with this, must-say wonderful, command system. I have spent a lot of time reading every available documentation and code piece. If @Citymonstret wants to take a look and maybe give us tips to avoid "ugly" workarounds or handling, we would be glad!_

The whole pull request has been tested, not yet thoroughly, you can have a global overview, but it is not really ready.

If you have any question or remark, do not hesitate!

# General Command Management
I have opted to use the annotations extension of the cloud command system. In my opinion, this is better suited to the current brush format handling which is all done inside brush classes.

**Executors have been kept and brush & performer command are still handled inside their classes.**

- ``Snipe`` class has been extended for a usage as a commander, because FAVS relies on a lot on this class.
- ``CommandRegistry`` is the main place for the whole handling behind the scenes. Otherwise, commands are registered as usually done in cloud, with some specific annotations when needed.

**``SniperCommander`` class is the commander to use with cloud command system. If the player exists, it returns its sniper. Otherwise, it returns a simple ``SniperSender``, similar to ``CommandSender``.**

# Command Manager
FAVS uses the paper command manager, when available, to enjoy some improvements. Falls back to bukkit command manager otherwise.

- Async tab-completions are enabled if available.
- ``Snipe``, ``PerformerSnipe`` & ``Toolkit`` classes are registered into the injector in order to be injected in command methods.
- Command exceptions are adapted and customized with the FAVS message syntax.

# Command Post-Processor
FAVS requires the command post-processor ins order handle specific FAVS behavior.

- Handles the ``@RequireToolkit`` annotation, makes sure the toolkit is available and the value stored.
- Handles the ``@DynamicRange`` annotation, used to define a range from non-constant variables, using reflection.
- Prepares the brush & performer when needed, their ``Snipe`` and stores them.

# Annotations & Parser
FAVS uses some annotations to facilitate development, based on common rules and behaviors.

- Handles the ``@RequireToolkit`` annotation, modifies the command meta.
- Handles the custom command execution method handler, which should differ for brush & performer. Cloud commands are designed to live in a class instance, this is not suitable to the current management of brush & performer. I have opted for a custom execution method which uses the brush & performer instance from the execution context instead of the base instance. _This avoid extra parameters for each command method._

# Arguments
FAVS needs a lot of custom arguments for either factories, registries, custom types, custom needs, etc.

**Suggestions & parsers are also declared via annotations in custom classes.**

# Other Changes
- Fixes #286
- Fixes #288
- Fixes #291
- Fixes #294
- Fixes #305
- Re-add ``meteor`` as an alias of the comet brush.
---
- All classes related to internal command managing from VS have been removed.
- ``FastAsyncVoxelSniper`` class has been removed. As far as I know, this class was useless and is now for sure.
- Some classes and methods have been added or refactored, but the overall code base is the same to try keeping maximum compatibility.
- Some translations have been reorganized or removed.
- Some code format has been modified, there will be another pull request next year hopefully to unify comments format.
- Improvements to brush properties loading. Previously, all aliases were loaded, subsequently loading the same brush several times.
- Modern switch syntax has replaced old ones.
- General improvements.

# Known Problems:
- There is currently one small issue with static/literal arguments and their aliases. Tab-completions are not handled for all aliases due to https://github.com/Incendo/cloud/blob/master/cloud-core/src/main/java/cloud/commandframework/arguments/StaticArgument.java#L134.
- Brigadier extension is voluntarily not used due to some incompatibilities with FAVS commands syntax.

```[tasklist]
### Submitter Checklist
- [x] Make sure you are opening from a topic branch (**/feature/fix/docs/ branch** (right side)) and not your main branch.
- [x] Ensure that the pull request title represents the desired changelog entry.
- [x] New public fields and methods are annotated with `@since TODO`.
- [x] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/.github/blob/main/CONTRIBUTING.md).
```

Aurelien30000 committed:
Cloud Command System Migration

# Introduction

**This is the first pass of the cloud command migration for FAVS.**
There will be a second one to restore old-fashioned command syntax, tracked as https://github.com/IntellectualSites/fastasyncvoxelsniper/issues/81.

_I have to highlight that this is my first experience with this, must-say wonderful, command system. I have spent a lot of time reading every available documentation and code piece. If @Citymonstret want to take a look and maybe give us tips to avoid "ugly" workarounds or handling, we would be glad!_

The whole pull request has been tested, not yet thoroughly, you can have a global overview, but it is not really ready.

If you have any question or remark, do not hesitate!

--

# General Command Management
I have opted to use the annotations extension of the cloud command system. In my opinion, this is better suited to the current brush format handling which is all done inside brush classes.

**Executors have been kept and brush & performer command are still handled inside their classes.**

- ``Snipe`` class has been extended for a usage as a commander, because FAVS relies on a lot on this class.
- ``CommandRegistry`` is the main place for the whole handling behind the scenes. Otherwise, commands are registered as usually done in cloud, with some specific annotations when needed.

**``SniperCommander`` class is the commander to use with cloud command system. If the player exists, it returns its sniper. Otherwise, it returns a simple ``SniperSender``, similar to ``CommandSender``.**

# Command Manager
FAVS uses the paper command manager, when available, to enjoy some improvements. Falls back to bukkit command manager otherwise.

- Async tab-completions are enabled if available.
- ``Snipe``, ``PerformerSnipe`` & ``Toolkit`` classes are registered into the injector in order to be injected in command methods.
- Command exceptions are adapted and customized with the FAVS message syntax.

# Command Post-Processor
FAVS requires the command post-processor ins order handle specific FAVS behavior.

- Handles the ``@RequireToolkit`` annotation, makes sure the toolkit is available and the value stored.
- Handles the ``@DynamicRange`` annotation, used to define a range from non-constant variables, using reflection.
- Prepares the brush & performer when needed, their ``Snipe`` and stores them.

# Annotations & Parser
FAVS uses some annotations to facilitate development, based on common rules and behaviors.

- Handles the ``@RequireToolkit`` annotation, modifies the command meta.
- Handles the custom command execution method handler, which should differ for brush & performer. Cloud commands are designed to live in a class instance, this is not suitable to the current management of brush & performer. I have opted for a custom execution method which uses the brush & performer instance from the execution context instead of the base instance. _This avoid extra parameters for each command method._

# Arguments
FAVS needs a lot of custom arguments for either factories, registries, custom types, custom needs, etc.

**Suggestions & parsers are also declared via annotations in custom classes.**

# Other Changes
- All classes related to internal command managing from VS have been removed.
- ``FastAsyncVoxelSniper`` class has been removed. As far as I know, this class was useless and is now for sure.
- Some classes and methods have been added or refactored, but the overall codebase is the same to try keeping maximum compatibility.
- Some translations have been reorganized or removed.
- Some code format has been modified, there will be another pull request next year hopefully to unify comments format.
- Improvements to brush properties loading. Previously, all aliases were loaded, subsequently loading the same brush several times.
- Modern switch syntax has replaced old ones.
- General improvements.

# Known Problems:
- There is currently one small issue with static/literal arguments and their aliases. Tab-completions are not handled for all aliases due to https://github.com/Incendo/cloud/blob/master/cloud-core/src/main/java/cloud/commandframework/arguments/StaticArgument.java#L134.
- Brigadier extension is voluntarily not used due to some incompatibilities with FAVS commands syntax.

NotMyFault commented on build.gradle.kts:
```diff
@@ -98,6 +104,7 @@ tasks {
         this.archiveFileName.set("${project.name}-${project.version}.${this.archiveExtension.getOrElse("jar")}")
         this.destinationDirectory.set(rootProject.tasks.shadowJar.get().destinationDirectory.get())
+        relocate("cloud.commandframework", "com.thevoxelbox.voxelsniper.cloud")
         relocate("org.incendo.serverlib", "com.thevoxelbox.voxelsniper.serverlib")
```
Need to add
```kt
relocate("io.leangen.geantyref", "com.thevoxelbox.geantyref")
```
That dependency is added by cloud.

Aurelien30000 committed:
Also relocate ``io.leangen.geantyref`` (thanks to @NotMyFault)

Aurelien30000 committed:
Fix https://github.com/IntellectualSites/fastasyncvoxelsniper/issues/286
- Fix ``/favs debugpaste`` link (format & clicking).
- Move commands reloading logic via ``/favs reload`` (broken & no longer needed).
- Re-add ``meteor`` as an alias of the comet brush.

Aurelien30000 committed:
Fix stencil brushes info when no file specified.

Aurelien30000 committed:
Fix https://github.com/IntellectualSites/fastasyncvoxelsniper/issues/288
- Set max brush size to a new reasonable value (500).
- Inform the sniper about any exceptions (unexpected errors).

Aurelien30000 committed:
Small reformat

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/DomeBrush.java:
Somewhere over here there is a stackoverflow happening, I'll link this one in an issue

OneLiteFeather replied:
Actually, your latest commit with the setblock and stackoverflow fixed that!

OneLiteFeather replied:
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/d84e39d7-cf82-416b-bdfc-6ac730f87df9)
There is an info command but it is not allowed here to enter somehow

Aurelien30000 replied:
There is no ``info`` sub-command for the dome brush. If it currently present in latest FAVS release, this is a mistake :)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ExtrudeBrush.java:
As I tried in game, it is showing me that I need to select a block but I can't seem to. Also not sure if there is something broken or not good explained, it could be a parameter still missing in this code
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/281fc6bb-78f5-470d-97c0-77423a82de04)

Aurelien30000 replied:
The used blocks come from the so-called voxel list (``/vl``). I will clarify the message with by explicitly mentioning the voxel list instead of blocks, thanks!

Aurelien30000 committed:
Clarify voxel list messages

Aurelien30000 committed:
Refactor sniper info sending and error message in case of exception (only)

Aurelien30000 committed:
Clean imports

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/JockeyBrush.java:
Looks like true or false aren't working, but info 
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/830281ba-6f9d-40c7-9b0f-76b23c7c0492)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/LightningBrush.java:
The same issue is here, the info parameter is shown in tabcompletion but is not allowed to be entered

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/MoveBrush.java:
Maybe there is something missing here as I can tabcomplete x y and z but it does not tell me a value to set and an error if I do it this way
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/1efa74f2-4373-4fdc-83e7-c4c522daa624)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/OceanBrush.java:
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/3131f768-3618-4815-aa99-d898c1e482eb)
Some other weird behaviour happens here

OneLiteFeather replied:
Don't know why but it is always 12

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/PaintingBrush.java:
The info argument is also missing here, but can be tab completed with the error above

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/PullBrush.java:
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/82598c8b-57cf-485c-ac90-69b7a26b2988)
Not sure if this is right with the different values, is that in Voxelsniper as well?

Aurelien30000 replied:
Good catch, thanks!

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RandomErodeBrush.java:
There is also no info here in this command

Aurelien30000 committed:
Fix pull brush parameters

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RulerBrush.java:
```diff
@@ -23,46 +27,56 @@ public class RulerBrush extends AbstractBrush {
+    public void onBrushInfo(
+            final @NotNull Snipe snipe
+    ) {
+        super.onBrushInfoCommand(snipe, Caption.of("oxelsniper.brush.ruler.info"));
```
well there is a v missing here 
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/c374a71d-274f-46d9-8751-d57bfd4ba861)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ScannerBrush.java:
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/4c464496-a18b-4931-86a5-52ad267438cb)
This is acting randomly here, I don't know where it is from

Aurelien30000 committed:
Fix typos

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/SnowConeBrush.java:
```diff
@@ -1,18 +1,32 @@
+    }
+
     @Override
     public void handleArrowAction(Snipe snipe) {
```
Is this wanted so that it only works with gunpowder? It doesn't seem to work on arrow item yet. I know that this might be not related to the migration.

Aurelien30000 replied:
This behavior has been here for 10 years :shrug: _We could probably add the same behavior for the arrow, as well._

Aurelien30000 committed:
Implements https://github.com/IntellectualSites/fastasyncvoxelsniper/issues/291 with the new command system

Aurelien30000 committed:
Upgrade cloud to 1.8.4

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ScannerBrush.java:
What if you are at y > 64 and the depth d is locked at 64 to look for the block? Minecraft has a different y value since 1.17.1 as far as I remember. Heighest block is Y=320, lowest is -64 in the overworld, heighest is Y=256 for the nether or the end. d should limit at the lowest and heighest value of the world I guess

Aurelien30000 replied:
This is not related to cloud migration and can be addressed separately.  Please open an issue for it ;)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/canyon/CanyonBrush.java:
Cloud works but the feature is still not right implemented, also having magic numbers (we'll have an issue for that soon)
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/1dbb31b5-e160-4871-aeef-6774a8231b44)

Aurelien30000 replied:
Magic numbers :thinking: What do you mean?

OneLiteFeather replied:
> Magic numbers 🤔 What do you mean?

Magic numbers are numbers that are set for no visible reason like when you use a dice and roll. Not sure how -54 to 60 can be explained

Aurelien30000 replied:
I see, indeed I have no real idea of what they are supposed to bound. Those values have probably been kept from the original VS (hopefully) and are configurable.

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/canyon/CanyonSelectionBrush.java:
working fine but also same thing with the brush (it creates a chunky hole at default, might be good to put that in documentation too)

Aurelien30000 replied:
This is not related to cloud migration and can be addressed separately, please open an issue for it ;)

OneLiteFeather commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/CheckerVoxelDiscBrush.java:
The most weird thing about this is that I can't find documentation for this feature and the brush is not a round disc (looks like a chess board, but not related to the CLOUD command, just to let you know that)
![grafik](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/77929075/ef5d2014-bd5b-4f73-8f85-93ba4930e8c1)

Aurelien30000 replied:
Probably a new addition but undocumented from former FAVS versions :sweat_smile:

OneLiteFeather replied:
Might be worth it as an issue xD

Aurelien30000 replied:
Yes!

OneLiteFeather replied:
Let's do this one later on in a different PR with an issue for that :)

OneLiteFeather commented on src/main/resources/lang/strings.json:
```diff
@@ -228,12 +237,11 @@
     "voxelsniper.brush.sign-overwrite.info": "&3The arrow writes the internal line buffer to the target sign.\n&3The gunpowder reads the text of the target sign into the internal buffer.\n&3Colors can be used using the \"&\" symbol.\n&6Sign Overwrite Brush Parameters:\n&b/b sio 1 [set|toggle] (...) -- Sets the text of the first sign line. (e.g. /b sio 1 set Blah Blah | /b sio 1 toggle)\n&b/b sio  ...
```
Is this broken somehow? The formatting doesn't show up in green

Aurelien30000 replied:
What is broken here :thinking: 
![image](https://github.com/IntellectualSites/fastasyncvoxelsniper/assets/43724816/a1060df6-627e-4e44-ad80-6ce824426df8)

OneLiteFeather replied:
well that is weird, the highlighting was missing in the class but when it is shown up here right, we can ignore that

OneLiteFeather reviewed (commented):
Looked out for in game functionality of brushes and their new cloud commands and some small errors / spelling I can identify, but I can't know if the code is "right" 100% for this PR (I am not a well trained developer). Opened some issues regarding the functionality, the implementation of CLOUD looks "good" (no errors I found, just the ones that are in this review) from user view.

OneLiteFeather reviewed (approved):
My review is done, let's go :)

NotMyFault committed:
Merge branch 'main' into feature/cloud-command-system
